### PR TITLE
First unread message

### LINKED ISSUE #1189: opening chat at the first unread message
the chat should open at the first unread message, not at the end.

we changed that recently on android, https://github.com/deltachat/deltachat-android/pull/1641, and that feels much better overall.

cmp. #1032

### PR DESCRIPTION
**Pull request by cyBerta**
closes #1189

### CODE REVIEW

File: DcCore/DcCore/DC/Wrapper.swift [hunk 1]
```diff
@@ -249,6 +250,10 @@ public class DcContext {
         return DcArray(arrayPointer: dc_get_fresh_msgs(contextPointer))
     }
 
+    public func getFreshMessagesCount(chatId: Int) -> Int {
```
Comment (r10s) [inline DcCore/DcCore/DC/Wrapper.swift:253]: there is already `getUnreadMessages` that does exactly the same under another name :)

so, for clarity, in this pr, i'd just this function - the things may get complicated enough :) is a subsequent pr, we can rename the function.

**Review by r10s — COMMENTED** [review]
nice!

as the line-separator is a "whole" UIView - can we have sth. as the day-titles here? so a string saying "123 new messages" centered with transparent background?

**Comment by cyBerta** [comment]
> as the line-separator is a "whole" UIView - can we have sth. as the day-titles here? so a string saying "123 new messages" centered with transparent background?

Yes we can do that. But we would have to change the behavior. Right now the thin line is only an indicator which messages the user should already have seen before they entered the chat, without doing an assumptions about the amount of unread messages. If the user sends an additional message, the thin line stays, if another message comes in, too. It currently only disappears if the user deletes a message, because that operation makes it harder to track if the thin line needs to be moved or not. It also disappears if the user re-enters the chat of course. 

If we add the information about the amount of messages we would have to remove or adapt the indicator as soon as a new message comes in or as soon as the user wrote an additional message. Moreover I like the fact that this is only a very unintrusive visual hint that doesn't take any place - a decent design :).

File: DcCore/DcCore/DC/Wrapper.swift (hunk 1 above)
Comment (cyBerta) [reply DcCore/DcCore/DC/Wrapper.swift:253]: Ups, I haven't seen that. Thanks! I'll fix that.

**Comment by r10s** [comment]
> Moreover I like the fact that this is only a very unintrusive visual hint that doesn't take any place - a decent design :)

hm, "decent design" is also true for a headline - maybe even more as no new visual element is introduced. that also has the advantage, that no questions pop up as it is self-explaining; also that is already used on android, so, we can say, that this actually works ux-wise.

File: deltachat-ios/Chat/ChatViewController.swift [hunk 2]
```diff
@@ -537,6 +537,12 @@ class ChatViewController: UITableViewController {
                 cell.update(text: "ErrDaymarker")
             }
             return cell
+        } else if id == DC_MSG_ID_MARKER1 {
+            // unread messages marker
+            let cell = tableView.dequeueReusableCell(withIdentifier: "info", for: indexPath) as? InfoMessageCell ?? InfoMessageCell()
+            let freshMsgsCount = self.messageIds.count - (indexPath.row + 1)
```
Comment (r10s) [inline deltachat-ios/Chat/ChatViewController.swift:543]: +1 for calculating the number of unread messages that way - so the headline can never be wrong :)

**Review by r10s — COMMENTED** [review]
great pr, pretty clear and straight forward. the new-message-headline looks great!

i played around with it a bit:

- the new-message-headline stays when new messages come in, as the counter is also updated, this makes a lot of sense. great

- for outgoing messages, however, currently the new-message-headline also stays, probably because messageIds array is not updated. i think, here, it should be removed, esp. as the outgoing message is counted as "new" as well.  
(maybe the easiest and most-performant way to fix that is to just remove DC_MSG_ID_MARKER1 from messagesIds manually - maybe there is even a swift-array-function for that, otherwise, maybe create sth. as removeValueFromArray working on an [Int] array)

**Comment by cyBerta** [comment]
>     * for outgoing messages, however, currently the new-message-headline also stays, probably because messageIds array is not updated. i think, here, it should be removed, esp. as the outgoing message is counted as "new" as well.

I've adapted the PR accordingly

**Pull request merged**